### ISSUE TITLE
Converting Existing Project to Build2

### ISSUE THREAD

**Issue by cmilhaupt** [build2/build2#174]
Given the following project structure (lots of multi libraries past first, second, and third):
```
my_project_repo/
├── multi-lib
│   ├── first
│   │   ├── src
│   │   │   ├── a.cpp
│   │   │   └── a.hpp
│   │   └── test
│   │       └── test_a.cpp
│   ├── second
│   │   ├── src
│   │   │   ├── b.cpp
│   │   │   └── b.hpp
│   │   └── test
│   │       └── test_b.cpp
│   ├── third
│   │   ├── src
│   │   │   ├── c.cpp
│   │   │   └── c.hpp
│   │   └── test
│   │       └── test_c.cpp
│   └── repositories.manifest
└── README.md
```
I'd like to convert from using cmake to build2. I would expect the following to work:
```
$ pwd
/tmp/my_project_repo
$ bdep new -l c++,cpp -t empty,no-readme multi-lib
$ cd multi-lib
$ bdep new --package -l c++,cpp -t lib,no-readme,no-test,prefix=src,no-subdir first
```
But this gives me the following: `error: generated version header is not supported in this layout`. 

Question 1: what does the version header do specifically? Does this mean my library won't be versioned?

If I add the `no-version` type in, I'm able to add the packages first, second, and third and I get the following manifest file in `multi-lib` (which I think is right?):
```
: 1
location: first/
:
location: second/
:
location: third/
```
In-source building now works (changing multi-lib/*/buildfile to only build `{src/}` instead of `{*/ -build/} manifest` since I get errors about the test dir), but if I try to build out of source (found [here](https://www.reddit.com/r/cpp/comments/qxedrx/comment/hl9c8ka/?utm_source=share&utm_medium=web2x&context=3), is this not documented anywhere else?) with
```
b multi-lib/@multi-lib-out/
```
Everything still builds in source.

Question 2: Should out of source building be working here with my setup, or have I missed something?

Question 3: Is this the correct approach to convert my project from cmake to build2, or is there a better path for me to take?

**Comment by boris-kolpackov** [comment]
> Question 1: what does the version header do specifically?

Version header is auto-generated from the `.in` template and contains preprocessor macros defined to your library's version (and its individual components). You can get an idea by creating a library project with, say, `bdep new -l lib libfoo` and checking the `version.hxx.in` file. Because it is auto-generated, we only support it if you use `subdir` and thus include your headers as `<libfoo/version.hxx>` rather than `<version.hxx>` (so we can map things back).

> Does this mean my library won't be versioned?

No, the auto-generated version header is just a convenience. You can maintain one manually or choose not to provide one at all.

> Question 2: Should out of source building be working here with my setup, or have I missed something?

There are two parts to answering this: first, do you want to use the entire toolchain or do you (for some reason) want to limit yourself to only using the build system?

If you don't have a good reason to avoid the entire toolchain, then I would strongly recommend that you "reset" by reading through the toolchain introduction: https://build2.org/build2-toolchain/doc/build2-toolchain-intro.xhtml In a sense what you are trying to achieve (setup out of source builds, etc) is "wrapped" into more convenient, high-level commands that fit into the overall workflow. So unless there is a good reason, I don't see why not take advantage of them.

If for some reason you only want to stick to the build system: what you've created is three independent build system projects that are tied together with a "glue buildfile" (you can search for this term in the build system manual for more information on this concept). Glue buildfiles are provided for convenience of building "whatever is configured in the source" and cannot be used to create an out-of-source build. To achieve that you would need to specify each project individually:

```
b multi-lib/first/@multi-lib-out/first/ multi-lib/second/@multi-lib-out/second/ multi-lib/third/@multi-lib-out/third/ 
```

> Question 3: Is this the correct approach to convert my project from cmake to build2, or is there a better path for me to take?

I think overall it's correct except I would use the entire toolchain and in particular the higher-level project manager (`bdep`). I would also make sure to read through the introduction -- it discusses these (e.g., multi-package projects, out-of-source builds, etc) and many other relevant topics.

**Comment by cmilhaupt** [comment]
> I would use the entire toolchain

Did one of the steps I pasted imply I'm not using the toolchain? Using `bdep new` is the only option I see in the introduction for creating a project. Also `b project/@project-out` is the only syntax mentioned in the toolchain introduction and in the linked section 1.3 "Output Directories and Scopes" of the build system manual. 

> I would also make sure to read through the introduction

I've read the introduction. Several times now actually. As well as the build system manual. I tried using the `bdep init` command to initialize an out of source build but couldn't get it to work with the above multi-package configuration either. I also don't have a need for building the projcet with clang or other compilers in a multi-configuration fashion. Since I asked about correct approaches, and you didn't mention `bdep init` in your response, I'm assuming that's not what you were referring to in pointing me to other toolchain resources.

> Glue buildfiles ... cannot be used to create an out-of-source build

So is there any way to use build2 to setup a multi package project that can build out-of-source? As I said, I'm coming from cmake, so a similar workflow (I can't type out my 30+ libraries in my multi-package configuration every time I want to build the project) would be preferred but there's some flexibility in project structure if there's no direct translation to build2. All I'm looking for is if it's possible to have a build directory at the same level as `multi-lib` in my initial example that would mimic the source tree with shared libraries/binaries at their appropriate level.

**Comment by boris-kolpackov** [comment]
Yes, `bdep init` is what I had in mind (which always creates an out of source build). I suspect it didn't work for you because you have created the project in a subdirectory of your repository (if you want to use `bdep`, then `packages.manifest` should be in the root of your repository). Here is a setup that worked for me:

```
cd /tmp/my_project_repo
bdep new -l c++,cpp -t empty,no-readme

mkdir multi-lib
cd multi-lib/

bdep new --package -l c++,cpp -t lib,no-readme,no-tests,prefix=src,no-subdir,no-version first
bdep new --package -l c++,cpp -t lib,no-readme,no-tests,prefix=src,no-subdir,no-version second
bdep new --package -l c++,cpp -t lib,no-readme,no-tests,prefix=src,no-subdir,no-version third

cd ..
tree ./
./
├── multi-lib/
│   ├── first/
│   │   ├── build/
│   │   │   ├── bootstrap.build
│   │   │   ├── export.build
│   │   │   └── root.build
│   │   ├── src/
│   │   │   ├── buildfile
│   │   │   ├── export.hpp
│   │   │   ├── first.cpp
│   │   │   └── first.hpp
│   │   ├── buildfile
│   │   └── manifest
│   ├── second/
│   │   ├── build/
│   │   │   ├── bootstrap.build
│   │   │   ├── export.build
│   │   │   └── root.build
│   │   ├── src/
│   │   │   ├── buildfile
│   │   │   ├── export.hpp
│   │   │   ├── second.cpp
│   │   │   └── second.hpp
│   │   ├── buildfile
│   │   └── manifest
│   └── third/
│       ├── build/
│       │   ├── bootstrap.build
│       │   ├── export.build
│       │   └── root.build
│       ├── src/
│       │   ├── buildfile
│       │   ├── export.hpp
│       │   ├── third.cpp
│       │   └── third.hpp
│       ├── buildfile
│       └── manifest
├── buildfile
├── packages.manifest
└── repositories.manifest

bdep init -C multi-lib-out @default cc
b                                                             # builds everything in multi-lib-out/
```

Is this something you had in mind?
